Quest: How much do you have to pay to use this parking spot?

marekkrug opened the pull request:
This PR is a new Quest with the name "How much do you have to pay to use this parking spot?" and was described in the Issue #6651.

westnordost committed:
refactor localization of weekdays, months, times

westnordost committed:
remove some old code

westnordost committed:
add texts for times, weekdays, months selectors

westnordost committed:
add month select and weekday select dialogs

westnordost committed:
Merge branch 'master' into opening_hours

westnordost committed:
add tests, fix bugs

westnordost committed:
add time format elements class plus tests

westnordost committed:
TimePicker mockup

westnordost committed:
finish time picker

westnordost committed:
add generic CheckboxList, analogous to the RadioGroup

westnordost committed:
rename preview function

westnordost committed:
make use of checkbox list in weekday and months select dialog

westnordost committed:
forgot default param

westnordost committed:
blub

westnordost committed:
add function for holidayselector -> holidays

westnordost committed:
actually, easier like this

westnordost committed:
more progress...

westnordost committed:
more screwing around

westnordost committed:
make time selectors column deal with both time points and time spans (#6646)

westnordost committed:
handle off days, somewhat

westnordost committed:
off days

Commit:
initial commit for parking_charge (mainly copied other quest so far)

Commit:
already implemented some basic things for quest, deleted a few files I copied too many

westnordost committed:
work on the data model

westnordost committed:
work on the data model

westnordost committed:
workweek

westnordost committed:
workweek

westnordost committed:
remove some old code

Commit:
more or less working quest layout, added icon for quest

Commit:
removed custom xml file, Quest is now using the standard composeViewBase. added new Stings to strings.xml and removed old ones.

paulklie commented on app/src/androidMain/kotlin/de/westnordost/streetcomplete/quests/parking_charge/AddParkingChargeForm.kt:
```diff
@@ -0,0 +1,196 @@
+    }
+
+    /** This function should probably be improved in the Future because it only matches some countries. */
+    private fun getCurrencyForCountry(): String = when (countryInfo.countryCode) {
```
I wonder if we should move this to [country metadata](https://github.com/streetcomplete/countrymetadata)

paulklie reviewed (commented):
great work so far, here are some ideas / topics for discussion

westnordost commented on app/src/androidMain/kotlin/de/westnordost/streetcomplete/quests/parking_charge/AddParkingChargeForm.kt:
```diff
@@ -0,0 +1,196 @@
+            "CA" -> "CAD"
+            else -> "EUR"
+        }
+}
```
We actually collect such meta-information in a separate repo:

https://github.com/streetcomplete/countrymetadata

In StreetComplete, we then parse this information via the `CountryInfo` class.

Also, this is apparently https://en.wikipedia.org/wiki/ISO_4217#Active_codes , so maybe add a comment this.

So, it would be good if you created a PR that adds a file like `data/currency.yml` in that repo.

westnordost commented on app/src/androidMain/kotlin/de/westnordost/streetcomplete/quests/parking_charge/AddParkingChargeForm.kt:
```diff
@@ -0,0 +1,196 @@
+                    Row(
+                        modifier = Modifier.fillMaxWidth(),
+                        horizontalArrangement = Arrangement.spacedBy(8.dp)
+                    ) {
```
To select a time unit, I would very much prefer a dropdown button (aka pop-up button). I already implemented a generic composable like this, have a look in `de.westnordost.streetcomplete.ui.common.DropdownButton`. See the `MaxHeightForm` for an example how it is used.

westnordost replied:
Also, the dropdown that just presents 30min and 15min is a bit problematic. Already now I see in taginfo a plethora of other minutes defined. Maybe we'd really need another input field in which the user can define x day(s), x hour(s) or x minute(s).

(This **may** again make the input form too wide to conveniently fit into one line)

westnordost reviewed (commented):
Regarding UI design, I'd somewhat prefer a UI like this

```
[        ] EUR / [ hour ▼ ]
```
 
The currency simply being a text (unless there are *actually* several currency in use in a country??).
(Bonus: It would be nice if not the ISO currency id was used in display, but the actual currency symbol, e.g. €, $, £, ...)

The amount, being a decimal input field, you can use the generic composable I created `de.westnordost.streetcomplete.ui.common.input.DecimalInput` which handles decimal points locale-aware. I guess allowed would be at most 2 decimal digits?

Finally, in the issue it came up that actual pricing may vary depending on a number of factors which may confuse the user what to answer. An example from the discussion on the wiki page:
- 1st hour: 2$ / hour
- subsequent hours: 1$ / hour

Is there a reasonable hint text that could be added to explain what the user should answer in this case?

Commit:
changed quest_parking_charge_title

mnalis commented on app/src/androidMain/kotlin/de/westnordost/streetcomplete/quests/parking_charge/AddParkingChargeForm.kt:
```diff
@@ -0,0 +1,196 @@
+            "HU" -> "HUF"
+            "RO" -> "RON"
+            "BG" -> "BGN"
+            "HR" -> "HRK"
```
Actually, Croatia ("HR") switched to "EUR" too [back in 2023](https://en.wikipedia.org/wiki/Croatia_and_the_euro)

westnordost committed:
defaults for new weekdays, new times, new months

westnordost committed:
note if months/weekdays are not specified

Commit:
new horizontal task layout with dropdown

Commit:
new horizontal task layout with dropdown

Commit:
added check_date:charge logic, now using correct tag charge:conditional instead of charge:description

Commit:
changed folder name from parking_charge to charge

westnordost committed:
make compile again (lobotomize AddParkingFeeForm), fix tests etc...

westnordost committed:
update country metadata

westnordost committed:
add missing strings

westnordost committed:
some aligning, paddingh

paulklie commented on app/src/androidMain/kotlin/de/westnordost/streetcomplete/quests/charge/AddParkingCharge.kt:
```diff
@@ -0,0 +1,42 @@
+class AddParkingCharge : OsmFilterQuestType<ParkingChargeAnswer>(), AndroidQuest {
+
+    override val elementFilter = """
+        nodes, ways, relations with amenity = parking
```
Hmm, not all payed parking areas have vending machines though. Here in sweden alot just have a sign/qr code so you can pay with app / sms.

westnordost committed:
refactor to a single add button (untested)

westnordost committed:
looks good now, except time selection dialog

westnordost committed:
fix time dialog

westnordost committed:
fix crash

westnordost committed:
add comment

westnordost committed:
solve enabling months differently

westnordost committed:
add divider

westnordost committed:
force user to specify months for all rules if he specified months for one rule (fixes #6175)

westnordost committed:
WIP add parking fee form

westnordost committed:
implement parking fee

westnordost committed:
don't support opening hours with incomplete months at all; force user to always specify months (fixes #6175)

westnordost committed:
display text for null

westnordost committed:
a bit nicer layout

westnordost committed:
Merge branch 'master' into opening_hours

# Conflicts:
#	app/src/androidMain/assets/country_metadata/IR.yml

westnordost committed:
remove unused styling

westnordost committed:
correction

westnordost committed:
add comment

westnordost committed:
Merge branch 'master' into opening_hours

westnordost committed:
remove own CheckboxList, rename MultipleSelectGroup

westnordost committed:
fix bugs reported by paulklie

Commit:
removed duplicate heading, bigger font for input, removed unnecessary quest strings

Commit:
german translations

Commit:
new ChargeInput composable

Commit:
new hint proposal (maybe could be changed later on)

westnordost committed:
some comments etc

westnordost committed:
feature: better default guess for initial time in the picker

westnordost committed:
correct title

Commit:
currency formatting: Automatic detection of locales like symbol before or after amount, currency symbol or locale symbol from country of quest

westnordost committed:
add comments

marekkrug commented:
@westnordost Today, I've implemented the CurrencyFormatter for both iOS and Android (I didn't get to test it on iOS so far because I don't use any apple products but this is how far i've got, but on Android it works). 
The quest now correctly shows the currency either in front or behind of the amount input field and chooses the currency via the `countryInfo.countryCode` of the quest the user is currently working on.

<img width="4320" height="2400" alt="screenshots-sc-charge-quest" src="https://github.com/user-attachments/assets/e384601f-9731-428a-94af-e8d65876df5d" />

I've also increased the font size and added a quest hint that is just a proposal. Maybe you have an opinion about how to improve it.

I'm still unsure about the right-to-left language compatability, I will try to have a look into them in the next few days :)

westnordost committed:
put duration unit dropdown into own file

westnordost reviewed (commented):
Regarding the currency formatter, it looks good on first view, I prefer an API like this, however:

1.

have an expect/actual 
```kt
expect class CurrencyFormatter(val locale: Locale) {
    fun format(amount: Double): String
    val currencyCode: String
}
```

2.
have a data class like this, which is in common code(!)
```kt
data class CurrencyFormatElements(
  val symbol: String,
  val isSymbolBeforeAmount: Boolean,
  val decimalDigits: Int,
) {
    companion object {
        fun of(locale: Locale?): CurrencyFormatElements {
          // TODO do the format of the test string, then regex on the result
        }
    }
}
```

Basically, similar structure/method as used here:

https://github.com/streetcomplete/StreetComplete/pull/6691/changes#diff-9f3a21c0597c3ee1c3d82960f150b9c6f0f876b19fbe67becf1a59789cfd5e69 (if the link does not correctly point to the file, the file is TimeFormatElements.kt)

This is mainly 
- consistent with the approach linked
- no need to implement that regex stuff on every platform

On whether this `CurrencyFormatElements.of(locale)` is the nicest pattern/approach, I don't have a strong opinion. It is just consistent to my approach with TimeFormatElements, but I could be swayed if there are reasons for having a different interface.

---

Also, don't forget to add unit tests for that!

westnordost commented:
Also, the dropdown for duration unit, I added 

https://github.com/streetcomplete/StreetComplete/pull/6691/changes#diff-5fcb8673a43b9e77b5d7bb3a8868616fb209fbf57d361a9b5bb818759b316520

Maybe you should base your PR on top of that PR (or wait until that is merged) so that you can just reuse that widget.

westnordost committed:
change string to be more generic

westnordost committed:
comment about ui thing

marekkrug committed:
Update app/src/androidMain/res/values/strings.xml

Co-authored-by: arthedain <73854742+paulklie@users.noreply.github.com>

Commit:
refactored CurrencyFormatter and introduced CurrencyFormatElements

Commit:
refactored CurrencyFormatter and introduced CurrencyFormatElements

Commit:
new tests for CurrencyFormatter and CurrencyFormatElements, added bit more Documentation

FloEdelmann committed:
Use typographic quotation marks and apostrophes

Commit:
migration to DurationUnit, so far not using DurationUnitDropdown

Commit:
Merge remote-tracking branch 'github/master'

# Conflicts:
#	app/src/androidMain/res/values/strings.xml

westnordost commented:
whats the problem with plural?

El 30 de enero de 2026 21:23:06 GMT+04:00, Marek Krug ***@***.***> escribió:
>marekkrug left a comment (streetcomplete/StreetComplete#6653)
>
>> Also, the dropdown for duration unit, I added
>> 
>> https://github.com/streetcomplete/StreetComplete/pull/6691/changes#diff-5fcb8673a43b9e77b5d7bb3a8868616fb209fbf57d361a9b5bb818759b316520
>> 
>> Maybe you should base your PR on top of that PR (or wait until that is merged) so that you can just reuse that widget.
>
>I've rebased the project to that branch and tried to use the DurationUnitDropdown. The problem is that all the values there are in plural ("hours", "minutes", "days"), but in OSM you need "hour", "minute", "day" and probably "night" and "day" in the Future (https://wiki.openstreetmap.org/wiki/Key:charge#Usage) 🤔
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/streetcomplete/StreetComplete/pull/6653#issuecomment-3824821971
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>

westnordost commented:
ah, right

El 31 de enero de 2026 2:59:22 GMT+04:00, Marek Krug ***@***.***> escribió:
>marekkrug left a comment (streetcomplete/StreetComplete#6653)
>
>writing `2.00 USD/hours` would be a bit weird in my opinion, in the wiki, the correct way to write it is `2.00 USD/hour`...
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/streetcomplete/StreetComplete/pull/6653#issuecomment-3826248774
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>

marekkrug committed:
Merge branch 'master' into master

Commit:
fixed quest number, now the app is in a runnable state again

westnordost commented:
I haven't been following this PR lately. Are all the previous comments resolved? As a PR author, you can tap that button yourself.

marekkrug commented:
I read through them and resolved a lot of them. Looks like there's nothing really remaining other than potential other quests.

westnordost committed:
add missing shadow (and make stack of dollars bigger)

westnordost commented on app/src/androidMain/kotlin/de/westnordost/streetcomplete/quests/charge/ParkingChargeAnswer.kt:
```diff
@@ -0,0 +1,16 @@
+    // e.g. "EUR"
+    val currency: String,
+    // either "day", "hour" or "minute"
+    val timeUnit: String
```
Unless there is a good reason not to do that, at least `amount` should be a `Double, `timeUnit` should be a `TimeUnit`.

westnordost commented on app/src/androidMain/kotlin/de/westnordost/streetcomplete/quests/charge/AddParkingChargeForm.kt:
```diff
@@ -0,0 +1,121 @@
+    DurationUnit.HOURS -> form.getString(R.string.quest_parking_charge_hour)
+    DurationUnit.DAYS -> form.getString(R.string.quest_parking_charge_day)
+    DurationUnit.MINUTES -> form.getString(R.string.quest_parking_charge_minute_short)
+}
```
Have a look at how the string resource is created for other quests / composable UI

westnordost replied:
👆

marekkrug replied:
Is this better now with the commit [fbb04b8](https://github.com/streetcomplete/StreetComplete/pull/6653/commits/fbb04b8792132aef6681427f9977d0e9b93bd894)? I tried to change the ChargeInput so that it now uses the DurationUnitDropdown but I'm not sure if I understood you correctly 🤔

westnordost replied:
Yes

westnordost reviewed (changes requested):
Phew, I am sorry to say this, but the currency formatter stuff is quite bonkers.

Each only in parts:

- the formatting is all over the place and there is commented out stuff plus open TODO comments left
- the doc comments make no sense
- the iOS implementation at least again in parts can not work. E.g. `getCurrencyCodeFromLocale` just completely ignores the passed parameter and looks like you copy-pasted the code from the other method
- the API of `CurrencyFormatter` (again, as for every point here in parts) make no sense (why is `getCurrencyCodeFromLocale` a method of an instance of `CurrencyFormatter`, what does the class have to do with a currency code?) 
- the tests for `CurrencyFormatter` only test half of the interface
- the implementation of `CurrencyFormatter` on both platforms is over-complex and also wrong. The parameter is a country code and you search through all available locales to find the first locale that uses the currency of the country with the given code in order to format it (because you need a locale to format). Now, imagine the Euro is used in a country where one formats currency like this "1.500,00 €" (France)  and in another country like this "€ 1 500.00" (Ireland). Oopsie! So, always the `Locale` (with region) must be passed to `CurrencyFormatter`.
- the second static function of `CurrencyFormatElements` looks like it is completely copy-paste from the other one, plus, only one of these is used (outside tests) anyway! Funnily enough, the one that is *not tested* is the one that is actually used, while the one that is tested is *not* used in production 🤦
- well, maybe more things, I lost the overview

I reckon that at least parts of this confusion might have been introduced as misunderstandings due to my requests to change your initial implementation of this, of course.
And since I looked into this code so deeply already, it makes sense if I rewrite that part myself rather than play PR ping pong. So, that's what I will do, you don't need to worry about this part.

---

The other parts I only skimmed through, but nevertheless I found some stuff. First and foremost, please make sure that you review your own code, too. I notice there are some TODOs and similar comments as well as sections of leftover commented out code.

For the other stuff, see below. Note however that this was just from skimming through it, there might be more in the details.

westnordost committed:
fix all things regarding the currency formatter

westnordost committed:
Merge branch 'master' into fork/marekkrug/master

# Conflicts:
#	app/src/androidMain/res/values-de/strings.xml

westnordost committed:
remove duplicate test

Commit:
changed SimpleCharge.amount to Double and SimpleCharge.timeUnit to DurationUnit. Also added formatting Option to DurationUnit

Commit:
added additional tests for Duration, temporarily added wrong title to quest to let it compile again for debugging

westnordost commented on app/src/commonMain/kotlin/de/westnordost/streetcomplete/osm/duration/Duration.kt:
```diff
@@ -4,15 +4,21 @@ import de.westnordost.streetcomplete.osm.duration.DurationUnit.*
-            MINUTES -> if (value != 1.0) "minutes" else "minute"
-            HOURS -> if (value != 1.0) "hours" else "hour"
-            DAYS -> if (value != 1.0) "days" else "day"
+    fun toOsmValue(alwaysSingular: Boolean = false, ignoreValue: Boolean = false): String {
```
Always singular? Ignore value? Looks like what you really want is to extend `DurationUnit` like this

```kt
enum class DurationUnit {
    MINUTES,
    HOURS,
    DAYS;

    fun toOsmValue(usePlural: Boolean = false): String = when (this) {
        MINUTES -> if (usePlural) "minutes" else "minute"
        HOURS -> if (usePlural) "hours" else "hour"
        DAYS -> if (usePlural) "days" else "day"
    } 
}
```

and use that accordingly, ditch `alwaysSingular: Boolean = false, ignoreValue: Boolean = false`

marekkrug replied:
But we never have the situation that the quest is with more than one time unit (`2 EUR/4 hours` would never happen). The quest is designed so that only one time unit is possible (`0.5 EUR/hour`).
Or am I missing something?

westnordost replied:
I don't think you are missing something, it's fine.

westnordost commented on app/src/androidMain/kotlin/de/westnordost/streetcomplete/quests/charge/ParkingChargeAnswer.kt:
```diff
@@ -0,0 +1,18 @@
+
+data class ItVaries(
+    val conditional: String
+) : ParkingChargeAnswer
```
Letting the user just input plain text for `charge:conditional`, e.g. `charge:conditional=on weekdays` is not acceptable.

This option must be removed or real alternative be implemented. Looking at current values for `charge:conditional`, I'd think this is not worth it, though. Better to leave a note.

westnordost commented on app/src/androidMain/kotlin/de/westnordost/streetcomplete/quests/charge/ChargeInput.kt:
```diff
@@ -0,0 +1,132 @@
+                    else -> "1.50"
+                }
+                Text(placeholderValue)
+            },
```
But why? Also, did you notice that this is not really localized, as the decimal point differs per language. Better just remove it.

westnordost reviewed (changes requested):
There are still some things commented out, and several TODOs. Do you need input for something?

Commit:
fixed durationUnit and duration, removed non-localized placeholder values

Commit:
updated test and fixed timeUnit usage

Commit:
removed all code for itVaries option (Only Issue creation is now allowed as other option from now on)

Commit:
Add parking charge quest

Commit:
refactor: adapt AddParkingCharge and AddParkingChargeForm to CommonMain

Commit:
new: added highlighting for other parking areas

Commit:
refractor: renaming stuff from ParkingCharge to Charge

Commit:
fix: improvements to CurrencyAmount, added Test coverage for CurrencyAmount

Commit:
fix: more Test coverage for CurrencyAmount

Commit:
fix: wrong conversion from CurrencyAmount to String in AddChargeForm

marekkrug commented:
@paulklie ready for review :)

marekkrug commented:
@paulklie ready for review :)

westnordost commented on app/src/commonMain/kotlin/de/westnordost/streetcomplete/quests/charge/AddChargeForm.kt:
```diff
@@ -0,0 +1,64 @@
+        isComplete = amount.let { it != null && (it.units != 0L || it.subunits != 0L) },
+        onClickOk = {
+            val amount = amount!!
+            val currency = CurrencyFormatter(countryInfo.userPreferredLocale).currencyCode ?: "???"
```
```suggestion
            val currency = CurrencyFormatter(countryInfo.userPreferredLocale).currencyCode ?: "¤"
```
(This is the generic/placeholder currency symbol)

westnordost commented on app/src/commonMain/kotlin/de/westnordost/streetcomplete/quests/charge/AddChargeForm.kt:
```diff
@@ -0,0 +1,64 @@
+
+    QuestForm(
+        on = on,
+        isComplete = amount.let { it != null && (it.units != 0L || it.subunits != 0L) },
```
```suggestion
        isComplete = amount != null && amount.units != 0 && amount.subunits != 0,
```

I just think this reads better, no functional change.

westnordost commented on app/src/commonMain/kotlin/de/westnordost/streetcomplete/osm/duration/Duration.kt:
```diff
@@ -18,5 +18,11 @@ data class Duration(val value: Double, val unit: DurationUnit) {
 enum class DurationUnit {
```
```suggestion
@Serializable
enum class DurationUnit {
```

westnordost commented on app/src/commonMain/kotlin/de/westnordost/streetcomplete/quests/charge/ChargeAnswer.kt:
```diff
@@ -0,0 +1,18 @@
+
+import de.westnordost.streetcomplete.osm.duration.DurationUnit
+
+sealed interface ChargeAnswer
```
This interface is not necessary when there is only one class that implements it.

Rename `SimpleCharge` to `Charge` and use that as the class for `AddParkingCharge`.

Don't forget to rename the file.

westnordost commented on app/src/commonMain/kotlin/de/westnordost/streetcomplete/util/locale/CurrencyAmount.kt:
This class (and the accompanying test) doesn't make sense to me.

Why not use a `Float` (or `Double` 🤷)? `Float` has enough precision to represent 2 decimals.

marekkrug replied:
It was a mix of

- It is nice to have a toString() method that does all the formatting
- It's nice to have it in a half-parsed form for converting it to a string later and to do logic with the units/subunits like this:
`isComplete = amount != null && (amount!!.units != 0L || amount!!.subunits != 0L),`
- I heard somewhere that, as a principle, you should never store money as a floating point number 
 
But I see the argument that it could be done a lot simpler without any intermediate structure. Should I refractor it to only use `Double`?

westnordost replied:
Yes please. As you want to format the string to have only X decimals and exactly this number of decimals, you could add a function like

```kt
/** Formats this double to exactly the given number of [decimals]. 
  * E.g. both `12.305` and `12.3` become `"12.30"` when [decimals] is 2. */
fun Double.formatPadded(decimals: Int): String {
    val int = toInt()
    val fraction = ((this - toInt()) * 10.0.pow(decimals)).toInt()
    val fractionPadded = fraction.toString().padStart(decimals, '0')
    return "$int.$fractionPadded"
}
```

to `de.westnordost.streetcomplete.util.ktx.Double.kt`

marekkrug replied:
done

marekkrug replied:
oh, I just found an error: if you input `2.4`, it tags `2.39`, the same with `2.8` -> `2.79`. Seems to be a rounding error...

marekkrug replied:
replacing `.toInt()` with `.roundToInt()` seems to fix it for now

westnordost commented on app/src/commonMain/kotlin/de/westnordost/streetcomplete/quests/charge/ChargeInput.kt:
```diff
@@ -0,0 +1,96 @@
+        )
+        Column(
+            verticalArrangement = Arrangement.spacedBy(4.dp)
+        ) {
```
what is this column for?

marekkrug replied:
was not needed any more

westnordost commented on app/src/commonMain/kotlin/de/westnordost/streetcomplete/quests/charge/ChargeInput.kt:
```diff
@@ -0,0 +1,96 @@
+                style = MaterialTheme.typography.h5,
+                color = MaterialTheme.colors.onSurface.copy(alpha = 0.6f),
+                modifier = Modifier.padding(end = 8.dp)
+            )
```
no need to add padding for each element, that is what `horizontalArrangement = Arrangement.spacedBy(8.dp)` in the `Row` is for.

Also, this is not a heading, so it shouldn't be styled as a heading. Also, usually only hints and similar get an alpha.

marekkrug replied:
What do style should I do for the heading? I did this styling [a while ago](https://github.com/streetcomplete/StreetComplete/pull/6653#discussion_r2643063719) and haven't changed it since.

westnordost replied:
I'd say, apply no style.

marekkrug replied:
<img width="749" height="232" alt="Screenshot_20260815_181046_StreetComplete Dev" src="https://github.com/user-attachments/assets/c93902f4-d8b7-4228-b52f-15103f4fb69f" />
👍

westnordost commented on app/src/commonMain/composeResources/values/strings.xml:
```diff
@@ -1395,10 +1395,17 @@
+    <string name="quest_parking_charge_title">How much do you need to pay to park here?</string>
+    <string name="quest_parking_charge_time_unit_label">per</string>
+    <string name="quest_parking_charge_hint">"For simplicity’s sake, it is recommended that you enter the standard amount for different amounts in this quest. For example, if it costs €3 during the day and €1 at night, enter €3. This makes it easier to compare prices from the data later on."</string>
+
```
```suggestion
```
This string is unused as far as I can see. (For good reason, IMO it is confusing.)

marekkrug replied:
It is being used as a hint. I discussed the hint with @paulklie [a while ago](https://github.com/streetcomplete/StreetComplete/pull/6653#discussion_r2702884841) and that was our result, but we can also remove it if you think that's better.

westnordost replied:
Hm well, I find it quite confusing. Also, as an advice to correctly specify it, somewhat questionable. The wiki page doesn't mention that it would be fine to only specify the "standard charge" (whatever that may be) if the situation is more complex and it would be very un-OSM-like to suggest something like this.

https://wiki.openstreetmap.org/wiki/Key:charge

€3 during day and €1 during night would also be taggable (just not in the app, because no interface for that has been built). E.g.

`charge=3 EUR/hour`
`charge:conditional=1 EUR/hour @ (20:00-06:00)`

marekkrug replied:
Just an idea:
`<string name="quest_parking_charge_hint">"Enter the charge and the time unit it applies to, e.g. per hour or per day. If the charge varies, e.g. by time of day or vehicle type, this cannot be fully expressed here yet. Either enter the amount that most commonly applies or create an Issue if you want to."</string>`

westnordost replied:
yes, please do

westnordost commented on app/src/commonMain/kotlin/de/westnordost/streetcomplete/quests/charge/ChargeInput.kt:
```diff
@@ -0,0 +1,96 @@
+        AutoFitTextFieldFontSize(
+            value = amountAsDouble?.toString().orEmpty(),
+            modifier = Modifier.weight(1f, fill = false)
+        ) {
```
Why this?

marekkrug replied:
I also thought The combination of AutoFit and the fillMaxWidth Modifier was redundant. But when I removed the fillMaxWidth, it was still filling up the whole area. You did use the same AutoFit in e.g. LengthMetersInput 🤔
```kotlin
@Composable
fun LengthMetersInput(
    length: Length.Meters?,
    onChange: (Length.Meters?) -> Unit,
    maxMeterDigits: Pair<Int, Int>,
    modifier: Modifier = Modifier,
    locale: Locale = Locale.current,
    style: TextFieldStyle = TextFieldStyle.Filled,
) {
    Row(
        modifier = modifier,
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        AutoFitTextFieldFontSize(
            value = length?.meters?.toShortString().orEmpty(),
            modifier = Modifier.weight(1f)
        ) {
            DecimalInput(
                value = length?.meters,
                onValueChange = { meters ->
                    onChange(meters?.let { Length.Meters(it) })
                },
                modifier = Modifier.fillMaxWidth(),
                maxIntegerDigits = maxMeterDigits.first,
                maxFractionDigits = maxMeterDigits.second,
                locale = locale,
                isUnsigned = true,
                style = style,
            )
        }
        Text("m")
    }
}
```

marekkrug replied:
<img width="1440" height="2304" alt="Screenshot_20260815_163049_StreetComplete Dev" src="https://github.com/user-attachments/assets/ab2688b8-7bf1-467a-8cfa-532041111d41" />
without AutoFitTextFieldFontSize
<img width="1440" height="2304" alt="Screenshot_20260815_162125_StreetComplete Dev" src="https://github.com/user-attachments/assets/5a813dd9-8a10-462d-8dd2-3cf1c9212cf6" />
with AutoFitTextFieldFontSize

marekkrug replied:
Should I just keep it that way or do you have a better option?

westnordost replied:
Oh, I see. Well then, the form looks good this way, so let's keep it as is.

westnordost reviewed (changes requested):
Can you post a screenshot/video on how the interface looks at the end?

Commit:
refractor: rename + simplify Charge quest

Commit:
fix: ChargeInput: remove unnecessary paddings + Column

Commit:
fix: Currencies without subunits like JPY are now working again (they always have subunits=0L), removed fillMaxWidth Modifier

marekkrug commented:
<img width="4320" height="2304" alt="bitmap" src="https://github.com/user-attachments/assets/1b872dee-0cf0-4708-bfda-939b33d3d486" />
Here are some Screenshots from the current state of the quest :)

westnordost commented on app/src/commonMain/kotlin/de/westnordost/streetcomplete/quests/charge/AddParkingCharge.kt:
```diff
@@ -0,0 +1,54 @@
+        and (
+            !charge and !charge:conditional
+            or charge older today -18 months
+        )
```
This means it would be asked also on `charge:conditional=...` when it is old enough. But since the app doesn't offer a way to specify more complicated charge conditions, it should **never** show again if `charge:conditional is specified. 

I.e.


```suggestion
        and !charge:conditional
        and (!charge or charge older today -18 months)
```

Commit:
fix: removed heading style

marekkrug committed:
fix: update elementFilter

Co-authored-by: Tobias Zwick <newton@westnordost.de>

Commit:
fix: removed CurrencyAmount, now doing everything without wrapper type and just with Double values

Commit:
fix: rounding issues with formatPadded, added option to omit zero fractions (e.g. 3.00 -> 3)

Commit:
removed hint for quest

marekkrug commented:
What are the next steps to finishing this PR? Do you still want to test the quest, @paulklie ?

westnordost commented on app/src/commonMain/kotlin/de/westnordost/streetcomplete/osm/duration/DurationUnitDropdown.kt:
```diff
@@ -14,6 +14,7 @@ import org.jetbrains.compose.resources.stringResource
     selectedDuration: DurationUnit,
     onSelectedDuration: (DurationUnit) -> Unit,
+    alwaysSingular: Boolean = false,
     modifier: Modifier = Modifier
```
```suggestion
    modifier: Modifier = Modifier,
    alwaysSingular: Boolean = false,
```

westnordost commented on app/src/commonMain/kotlin/de/westnordost/streetcomplete/quests/charge/AddParkingCharge.kt:
```diff
@@ -0,0 +1,51 @@
+
+    override fun getHighlightedElements(element: Element, mapData: MapDataWithGeometry) =
+        mapData.filter("""
+             nodes, ways with amenity = parking
```
```suggestion
             nodes, ways, relations with amenity = parking
```

westnordost commented on app/src/commonMain/kotlin/de/westnordost/streetcomplete/quests/charge/ChargeInput.kt:
```diff
@@ -0,0 +1,80 @@
+        if (!currencyFormatInfo.isSymbolBeforeAmount) {
+            Text(
+                text = currencyFormatInfo.symbol,
+                style = MaterialTheme.typography.h5,
```
```suggestion
```

westnordost commented on app/src/commonMain/kotlin/de/westnordost/streetcomplete/quests/charge/AddChargeForm.kt:
```diff
@@ -0,0 +1,64 @@
+        isComplete = amount != null && amount != 0.0,
+        onClickOk = {
+            val amount = amount!!
+            val currency = CurrencyFormatter(countryInfo.userPreferredLocale).currencyCode ?: "¤"
```
```suggestion
            val currency = CurrencyFormatter(countryInfo.userPreferredLocale).currencyCode ?: "???"
```

Ah, it's the currency **code**, not the currency symbol. Then, actually, "???" was better

westnordost commented on app/src/commonMain/kotlin/de/westnordost/streetcomplete/quests/charge/AddChargeForm.kt:
```diff
@@ -0,0 +1,64 @@
+                durationUnit = durationUnit,
+                onDurationUnitChange = { durationUnit = it },
+                perLabel = stringResource(Res.string.quest_parking_charge_time_unit_label),
+                modifier = Modifier.padding(16.dp)
```
```suggestion
```

westnordost reviewed (approved):
Looks good to me now!

westnordost committed:
Apply suggestions from code review

Co-authored-by: Tobias Zwick <newton@westnordost.de>
